storage/remote/wal_watcher.go

```diff
@@ -0,0 +1,362 @@
+		series   []tsdb.RefSeries
+		samples  []tsdb.RefSample
+		nSeries  = 0
+		nSamples = 0
+	)
+	switch dec.Type(rec) {
+	case tsdb.RecordSeries:
+		series, err := dec.Series(rec, series[:0])
+		if err != nil {
+			level.Error(log.With(w.logger)).Log("err", err)
+		}
+		watcherSeriesRecordsRead.WithLabelValues(w.writer.Name()).Add(float64(len(series)))
+		w.writer.StoreSeries(series)
+		nSeries += len(series)
+	case tsdb.RecordSamples:
+		samples, err := dec.Samples(rec, samples[:0])
+		if err != nil {
+			level.Error(log.With(w.logger)).Log("err", err)
+			break
+		}
+		watcherSamplesRecordsRead.WithLabelValues(w.writer.Name()).Add(float64(len(samples)))
+		// Blocks  until the sample is sent to all remote write endpoints or closed.
+		w.writer.Append(samples)
+		level.Debug(w.logger).Log("msg", "wrote samples to queue manager", "samples", len(samples))
+
+		nSamples += len(samples)
+	case tsdb.RecordTombstones:
+		w.logger.Log("msg", "tombstone record")
+	case tsdb.RecordInvalid:
+		w.logger.Log("msg", "invalid record")
+	default:
+		level.Info(w.logger).Log("msg", "unknown TSDB record type in decodeSegment")
+	}
+
+}
+
+func (w *WALWatcher) readSegment(r *wal.LiveReader) {
+	for r.Next() == nil && !isClosed(w.quit) {
+		w.decodeRecord(r.Record())
+	}
+}
+
+func (w *WALWatcher) watch(wl *wal.WAL, segment *wal.Segment, reader *wal.LiveReader) error {
+	segmentTimeout := 100 * time.Millisecond
+	segmentTicker := time.NewTicker(segmentTimeout)
+	defer segmentTicker.Stop()
+	dirTicker := time.NewTicker(1 * time.Second)
+	defer dirTicker.Stop()
+
+	currentSegmentName := fmt.Sprintf("%08d", w.currentSegment)
+	currentSegmentMetric.WithLabelValues(w.writer.Name()).Set(float64(w.currentSegment))
+
+	defer segment.Close()
+
+	for {
+		select {
+		case <-w.quit:
+			level.Info(w.logger).Log("msg", "quitting WAL watcher watch loop")
+			return errors.New("quit channel")
+		case <-dirTicker.C:
+			// check if there is a new checkpoint
+			dir, _, err := tsdb.LastCheckpoint(w.walDir)
+			if err != nil && err != tsdb.ErrNotFound {
+
+			}
+			if err == nil && dir != w.lastCheckpoint {
+				level.Info(w.logger).Log("msg", "new checkpoint detected", "last", w.lastCheckpoint, "new", dir)
+				w.lastCheckpoint = dir
+				err = w.readCheckpoint(path.Join(w.walDir, w.lastCheckpoint))
+				if err != nil {
+					w.logger.Log("err", err)
+				}
+			}
+			// check if new segments exist
+			_, last, err := wl.Segments()
+			if err != nil {
+				level.Error(w.logger).Log("err", err)
+				continue
+			}
+			if last > w.currentSegment {
+				w.logger.Log("msg", "read segment", "segment", fmt.Sprintf("%+v", segment))
+				w.readSegment(reader)
+				w.logger.Log("msg", "a new segment exists, we should start reading it", "current", currentSegmentName, "new", fmt.Sprintf("%08d", last))
+				return nil
+			}
+		case <-segmentTicker.C:
+			w.readSegment(reader)
+		}
+	}
+}
+
+func (w *WALWatcher) runWatcher() {
+	// The WAL dir may not exist when Prometheus first starts up.
```

Comment:
You could do this with one for loop

Reply:
Not sure what you're saying here; do the entire function in a single for loop?